operators/src/upgrade/rest/service.rs

```diff
@@ -125,9 +138,179 @@ pub async fn apply_upgrade() -> Result<HttpResponse, RestError> {
     }
 }
 
+pub async fn upgrade_data_plane() -> Result<(), Error> {
+    let pods: Api<Pod> = Api::namespaced(
+        UpgradeConfig::get_config().k8s_client().client(),
+        UpgradeConfig::get_config().namespace(),
+    );
+
+    let io_engine_listparam = ListParams::default().labels(IO_ENGINE_POD_LABEL);
+    let initial_io_engine_pod_list: ObjectList<Pod> = pods.list(&io_engine_listparam).await?;
+    for pod in initial_io_engine_pod_list.iter() {
+        // Fetch the node name on which the io-engine pod is running
+        let node_name = pod
+            .spec
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("pod.spec.node_name is invalid"))?
+            .node_name
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("pod.spec.node_name is invalid"))?
+            .as_str();
+
+        tracing::info!(
+            "Upgrade starting for data-plane pod: {:#?} present on node: {:#?} ",
+            pod.metadata.name,
+            node_name
+        );
+
+        // Drain the node
+        match UpgradeConfig::get_config()
+            .rest_client()
+            .nodes_api()
+            .put_node_drain(node_name, DRAIN_FOR_UPGRADE)
+            .await
+        {
+            Ok(_) => {
+                // Wait for any node drain process to complete across the cluster.
+                while is_draining().await? {
+                    tokio::time::sleep(Duration::from_secs(10_u64)).await;
+                }
+                tracing::info!("Mayastor node: {:#?} drained successfully", node_name);
+
+                // Wait for 30 seconds for any rebuilds to kick in.
+                tokio::time::sleep(Duration::from_secs(30_u64)).await;
+
+                // wait for the rebuild to cpmplete if any
+                while is_rebuilding().await? {
+                    tokio::time::sleep(Duration::from_secs(10_u64)).await;
+                }
+
+                // Deleting the io-engine pod
+                let pod_name = pod.metadata.name.as_ref().unwrap();
+                tracing::info!(
+                    "Deleting io-engine pod: {:#?} present on node: {:#?}",
+                    pod_name,
+                    node_name,
+                );
+
+                pods.delete(pod_name, &DeleteParams::default())
+                    .await?
+                    .map_left(|o| tracing::debug!("Deleting Pod: {:?}", o.status))
+                    .map_right(|s| tracing::info!("Deleted Pod: {:?}", s));
```

Comment:
If result is `Either::Left`, do we need to wait for deleted?

Reply:
no, we dont ned to wait, we just need to issue the delete command on pod.